crates/sui-framework/sources/governance/voting_power.move

```diff
@@ -0,0 +1,256 @@
+    /// power less than the max.
+    /// Similarly, "leftover" voting power due to rounding error is distributed *equally* among validators with voting power less than the max (if possible),
+    /// and among all validators if everyone is already at the max.
+    /// This function ensures the following invariants:
+    /// 1. Total voting power of all validators sums to `TOTAL_VOTING_POWER`
+    /// 2. `active_validators` is sorted by voting power in descending order
+    /// 3. `active_validators` is sorted by stake in descending order
+    /// This function attempts to maintain the following invariants whenever possible:
+    /// 4. Each validator's voting power is <= `MAX_VOTING_POWER`
+    /// 5. If validator A and B have the same stake, they will have the same voting power
+    /// Invariant (4) and (5) should almost always hold for Sui in practice due to high validator count, and stakes that aren't exactly
+    /// equal, but in theory/in tests these can be violated due to:
+    /// - a staking distribution like [1, 1, 1, 1] (will violate (4))
+    /// - a staking distribution where at least one validator has `MAX_VOTING_POWER`, and there is a remainder after redistributing the
+    ///   leftovers due to odd numbers. In this case, a single validator will have 1 more than its proportionsal share
+    public fun update(active_validators: &mut vector<Validator>): u64 {
+        // sort validators by stake, in descending order
+        bubble_sort_by_stake(active_validators);
+
+        let total_stake = total_stake(active_validators);
+        let voting_power_remaining = TOTAL_VOTING_POWER;
+        let last_voting_power_remaining = voting_power_remaining;
+        let first_non_max_idx = 0;
+        let num_validators = vector::length(active_validators);
+
+        // zero out voting power
+        let i = 0;
+        while (i < num_validators) {
+            let validator = vector::borrow_mut(active_validators, i);
+            validator::set_voting_power(validator, 0);
+            i = i + 1;
+        };
+
+        loop {
+            let i = first_non_max_idx;
+            // distribute voting power proportional to stake, but capping at MAX_VOTING_POWER
+            while (i < num_validators) {
+                let validator = vector::borrow_mut(active_validators, i);
+                let validator_stake = validator::total_stake(validator);
+                let prev_voting_power = validator::voting_power(validator);
+                let voting_power_share = (last_voting_power_remaining * validator_stake) / total_stake;
+                let new_voting_power = prev_voting_power + voting_power_share;
+                let voting_power_distributed = if (new_voting_power >= MAX_VOTING_POWER) {
+                    validator::set_voting_power(validator, MAX_VOTING_POWER);
+                    first_non_max_idx = i + 1;
```

Comment:
If we continue updating this in the loop, I don't understand why it's called first_non_max_idx

Reply:
I added some comments on this.